nixos/modules/virtualisation/nixos-containers.nix

```diff
@@ -939,7 +940,15 @@ in
             optional (cfg.networkNamespace != null && (cfg.privateNetwork || cfg.interfaces != [ ]))
               "containers.${name}.networkNamespace is mutally exclusive to containers.${name}.privateNetwork and containers.${name}.interfaces.";
         in
-        mkMerge (mapAttrsToList mapper config.containers);
+        [
+          (mkMerge (mapAttrsToList mapper config.containers))
+          {
+            assertion =
+              !config.boot.enableContainers
+              -> !(any (lib.strings.hasPrefix "container@") (attrNames config.systemd.services));
```

Comment:
If we’re asserting on this I think we need to include the `systemd.targets.*.wants` form too. As I said though, I think this can be dropped, as the use case described involved manually using the command‐line tool, which this wouldn’t catch – if we wanted friendlier messaging around that, we could install a stub `nixos-container` tool that prints an error message when they’re disabled, but I hope that the release note should be sufficient here, as the tool going missing will be obvious.